V0.45.0 release notes (#3115)
grafana/k6: release notes/v0.45.0.md

  stream.write({
    lo: {
      latitude: 400000000,
      longitude: -750000000,
    },
    hi: {
      latitude: 420000000,
      longitude: -730000000,
    },
  });

  sleep(0.5);
};
```
</details>

You can just replace `k6/grpc` import with `k6/experimental/grpc` to use the new functionality.

In the future this functionality will be moved to the `k6/grpc` module.

### You can now only upload a test to the cloud without running it [#3030](https://github.com/grafana/k6/pull/3030)

For [years users have wanted](https://github.com/grafana/k6-cloud-feature-requests/issues/22) to be able to update the test that is saved in the cloud but *not* run it at this exact point.

This is now possible by adding `--upload-only` when invoking `k6 cloud` as in `k6 cloud --upload-only script.js`.

This is likely going to be most useful in a CI on the actual test script project. Now that CI can just run `k6 cloud --upload-only new-verson-of-script.js` on "release".

And later on that newer version will be used. For example by a [scheduled run](https://k6.io/docs/cloud/manage/scheduled-tests/). 

### Setting sample metadata API [#3037](https://github.com/grafana/k6/pull/3037)

> Support for high-cardinality metrics metadata was added in v0.41.0, but it wasn't accessible from test scripts. It's now possible to set or delete metadata with a similar API as used for tags:

Author: ```suggestion
Support for high-cardinality metrics metadata was added in v0.41.0, but it wasn't accessible from test scripts. It's now possible to set or delete metadata for the whole VU with a similar API as used for tags:
```

k6 v0.45.0 is here 🎉! This release includes:

- Experimental gRPC streaming support.
- Update scripts in the cloud without running tests.
- JS Metadata API.
- A lot of internal changes and bugfixes.

## Breaking changes

> * [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names incompatible with OpenTelemetry or Prometheus. This is planned to become an error with v0.48.0. 

Reviewer: ```suggestion
* [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names incompatible with OpenTelemetry or Prometheus. This is planned to [become an error](https://github.com/grafana/k6/issues/3065) with v0.48.0.
```

> 

Reviewer 2: Add k6 browser breaking changes:
```suggestion
* [browser#851](https://github.com/grafana/xk6-browser/pull/851) Remove existing browser namespaced metrics. These have been deprecated in favor of Web Vitals metrics.

```

}
```

This also introduces the sub-object `metrics` on the `vu` object.
Apart from `metadata` it has another property `tags`. This is meant to be the new way to set tags instead of using `exec.vu.tags`. 

There are no current plans to replace `exec.vu.tags` with `exec.vu.metrics.tags`.

### UX improvements and enhancements

- [#3015](https://github.com/grafana/k6/pull/3015) We're now building ARM64 Docker images in CI, but due to some Docker complications, they're still not pushed to the repository. We expect to resolve these issues by next release. Thanks @nrxr! 🙇
- [#3099](https://github.com/grafana/k6/pull/3099) replace "breached" with "crossed" in logs around thresholds. Thanks to @MattDodsonEnglish 🙇.
- [#3102](https://github.com/grafana/k6/pull/3102) Better error message when SharedArray constructor is provided with an async function. This is not supported, but the original message wasn't very clear.
- [#3089](https://github.com/grafana/k6/pull/3089) Add Software Bill of Materials (SBOM) reports to k6 releases. Thanks to @SadFaceSmith 🙇.

## Bug fixes

* [#3058](https://github.com/grafana/k6/pull/3058) fix repetitive `the`. Thank you, @cuishuang 🙇.

> 

Reviewer 2: Add k6 browser bug fixes:
```suggestion
* [browser#852](https://github.com/grafana/xk6-browser/pull/852) Fix `Locator.WaitFor` for `detached` and `hidden` states.
* [browser#859](https://github.com/grafana/xk6-browser/pull/859) Fix remote object parsing when subtype is `null`.

```

k6 v0.45.0 is here 🎉! This release includes:

- Experimental gRPC streaming support.
- Update scripts in the cloud without running tests.
- JS Metadata API.
- A lot of internal changes and bugfixes.

## Breaking changes

* [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names incompatible with OpenTelemetry or Prometheus. This is planned to [become an error](https://github.com/grafana/k6/issues/3065) with v0.48.0.
* [browser#851](https://github.com/grafana/xk6-browser/pull/851) Remove existing browser namespaced metrics. These have been deprecated in favor of Web Vitals metrics.

## New features

### Experimental gRPC module with streaming support [#3107](https://github.com/grafana/k6/pull/3107)

> There is a new experimental module `k6/experimental/grpc`. It is a copy of the `k6/grpc` module with added stream support [#2020](https://github.com/grafana/k6/issues/2020). 

Reviewer 3: ```suggestion
There is a new experimental module `k6/experimental/grpc`. It is a copy of the `k6/net/grpc` module with added stream support [#2020](https://github.com/grafana/k6/issues/2020). 
```

// (golang should be installed)
const GRPC_ADDR = __ENV.GRPC_ADDR || '127.0.0.1:10000';
const GRPC_PROTO_PATH = __ENV.GRPC_PROTO_PATH || '../../grpc_server/route_guide.proto';

let client = new Client();

client.load([], GRPC_PROTO_PATH);

export default () => {
  client.connect(GRPC_ADDR, { plaintext: true });

  const stream = new Stream(client, 'main.FeatureExplorer/ListFeatures', null);

  stream.on('data', function (feature) {
    console.log(
      'Found feature called "' +
        feature.name +
        '" at ' +
        feature.location.latitude / COORD_FACTOR +
        ', ' +
        feature.location.longitude / COORD_FACTOR
    );
  });

  stream.on('end', function () {
    // The server has finished sending
    client.close();
    console.log('All done');
  });

  stream.on('error', function (e) {
    // An error has occurred and the stream has been closed.
    console.log('Error: ' + JSON.stringify(e));
  });

  // send a message to the server
  stream.write({
    lo: {
      latitude: 400000000,
      longitude: -750000000,
    },
    hi: {
      latitude: 420000000,
      longitude: -730000000,
    },
  });

  sleep(0.5);
};
```
</details>

> You can just replace `k6/grpc` import with `k6/experimental/grpc` to use the new functionality.
> 
> In the future this functionality will be moved to the `k6/grpc` module.

Reviewer 3: ```suggestion
You can just replace `k6/grpc` import with `k6/experimental/grpc` to use the new functionality. Documentation for the module is available [here](https://k6.io/docs/javascript-api/k6-experimental/grpc/).

In the future, this functionality will be moved to the `k6/net/grpc` module.
```

### You can now only upload a test to the cloud without running it [#3030](https://github.com/grafana/k6/pull/3030)

For [years users have wanted](https://github.com/grafana/k6-cloud-feature-requests/issues/22) to be able to update the test that is saved in the cloud but *not* run it at this exact point.

This is now possible by adding `--upload-only` when invoking `k6 cloud` as in `k6 cloud --upload-only script.js`.

> This is likely going to be most useful in a CI on the actual test script project. Now that CI can just run `k6 cloud --upload-only new-verson-of-script.js` on "release".

Reviewer 4: ```suggestion
This is likely going to be most useful in a CI on the actual test script project. Now that CI can just run `k6 cloud --upload-only new-version-of-script.js` on "release".
```

And later on that newer version will be used. For example by a [scheduled run](https://k6.io/docs/cloud/manage/scheduled-tests/). 

### Setting sample metadata API [#3037](https://github.com/grafana/k6/pull/3037)

Support for high-cardinality metrics metadata was added in v0.41.0, but it wasn't accessible from test scripts. It's now possible to set or delete metadata for the whole VU with a similar API as used for tags:
```javascript
import exec from "k6/execution";

export default () => {

>     exec.vu.metrics.metadata["my_cool_id"] = "a very unique value"
>     // all metrics from here on will have this metadata set
>     delete exec.vu.metrics.metadata["my_cool_id"]
>     // all metriccs from here on wil *not* have the metadata set

Reviewer 4: ```suggestion
  exec.vu.metrics.metadata["my_cool_id"] = "a very unique value";
  // all metrics from here on will have this metadata set
  delete exec.vu.metrics.metadata["my_cool_id"];
  // all metrics from here on will *not* have the metadata set
```

}
```

This also introduces the sub-object `metrics` on the `vu` object.
Apart from `metadata` it has another property `tags`. This is meant to be the new way to set tags instead of using `exec.vu.tags`. 

There are no current plans to replace `exec.vu.tags` with `exec.vu.metrics.tags`.

### UX improvements and enhancements

- [#3015](https://github.com/grafana/k6/pull/3015) We're now building ARM64 Docker images in CI, but due to some Docker complications, they're still not pushed to the repository. We expect to resolve these issues by next release. Thanks @nrxr! 🙇
- [#3099](https://github.com/grafana/k6/pull/3099) replace "breached" with "crossed" in logs around thresholds. Thanks to @MattDodsonEnglish 🙇.
- [#3102](https://github.com/grafana/k6/pull/3102) Better error message when SharedArray constructor is provided with an async function. This is not supported, but the original message wasn't very clear.
- [#3089](https://github.com/grafana/k6/pull/3089) Add Software Bill of Materials (SBOM) reports to k6 releases. Thanks to @SadFaceSmith 🙇.

## Bug fixes

> * [#3058](https://github.com/grafana/k6/pull/3058) fix repetitive `the`. Thank you, @cuishuang 🙇.

Reviewer 4: This is not a bug fix. Let's move it to the "Maintenance and internal improvements" section?

* [browser#852](https://github.com/grafana/xk6-browser/pull/852) Fix `Locator.WaitFor` for `detached` and `hidden` states.
* [browser#859](https://github.com/grafana/xk6-browser/pull/859) Fix remote object parsing when subtype is `null`.

## Maintenance and internal improvements

> - [#2091](https://github.com/grafana/k6/pull/2991) Refactor JS modules system so that is usable in tests. Which allowed the enablement of the tc39 tests for modules [#3040](https://github.com/grafana/k6/pull/3040).

Reviewer 4: ```suggestion
- [#2091](https://github.com/grafana/k6/pull/2991) Refactor JS modules system so that is usable in tests. Which allowed enabling the tc39 tests for modules [#3040](https://github.com/grafana/k6/pull/3040).
```

- [#3078](https://github.com/grafana/k6/pull/3078) Pinning to Alpine 3.17 in Docker builder image. Thank you, @arukiidou 🙇.
- [#3086](https://github.com/grafana/k6/pull/3086) Fix downloading `.golangci.yml` for PRs from forks. 
- [#3088](https://github.com/grafana/k6/pull/3088) Make TestDNSResolver less flaky. 
- [#3094](https://github.com/grafana/k6/pull/3094) Fix example from the run command. Thanks to @rocktwotj 🙇.
- [#3095](https://github.com/grafana/k6/pull/3095) Maintenance update of `.golangci.yml`.
- [#3103](https://github.com/grafana/k6/pull/3103) Fix lint and logical issues in `k6/data` module tests.
- [#3045](https://github.com/grafana/k6/pull/3045), [#3049](https://github.com/grafana/k6/pull/3049), [#3073](https://github.com/grafana/k6/pull/3073) and [#3044](https://github.com/grafana/k6/pull/3044) New issues are now automatically assigned to maintainers, to improve response time on issues. Both new issue and new PR assignments are now not using external actions.
- [#3109](https://github.com/grafana/k6/pull/3109) Add a way to get the cloudapi Client's base URL. Thanks to @yorugac 🙇.
## Roadmap

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028)

Reviewer 4: ```suggestion
There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028).
```

The idea is that we will keep it updated and current or future users will be able to look at it to see on what bigger projects the team is working on.

### Cloud output v2

Work on a new version of the cloud output has been ongoing over this cycle.

While functionally it is now mostly complete, we feel like more testing is still needed and some smaller issues need to be ironed out.

> Ove the next cycle we will be testing it internally and coming v0.46.0 it will be generally available. With the idea of becoming the default.

Reviewer 5: ```suggestion
Over the next cycle we will be testing it internally and coming v0.46.0 it will be generally available. With the idea of becoming the default.
```

> Over the next cycle we will be testing it internally and coming v0.46.0 it will be generally available. With the idea of becoming the default.

Reviewer 4: We should clarify what will happen to the current version.

```suggestion
Over the next cycle we will be testing it internally and coming v0.46.0 it will be generally available, with the goal of it becoming the default Cloud output. It will still be possible to use the current version via an option, but we plan to gradually deprecate it.
```

@codebien Is this accurate?

Reviewer: My interpretation is that it means after we've changed the default option, so yes, it is accurate. Because it will be possible to fallback on v1 until we merge https://github.com/grafana/k6/pull/3092 (targeted as v0.48.0).

Reviewer 4: OK. I changed the first sentence to "Over the next cycle we will be testing it internally, and in v0.46.0 it will be generally available as the default Cloud output.".

The new output has some benefits over the previous one:

> - Binary(protobuf) format instead of JSON [#2963](https://github.com/grafana/k6/pull/2963)

Reviewer 4: ```suggestion
- Binary (protobuf) format instead of JSON [#2963](https://github.com/grafana/k6/pull/2963)
```

- Samples aggregation for every metric instead of only for HTTP ones [#3071](https://github.com/grafana/k6/pull/3071)
- HDR Histogram generation for trend-type metrics [#3027](https://github.com/grafana/k6/pull/3027)

> This in general makes the payload send for tests with a lot of samples *way* smaller which also in most cases has turned out to lower the CPU and memory usage.

Reviewer 4: ```suggestion
This in general makes the payload sent for tests with a lot of samples much smaller, which also in most cases has turned out to lower the CPU and memory usage.
```

k6 v0.45.0 is here 🎉! This release includes:

- Experimental gRPC streaming support.
- Update scripts in the cloud without running tests.
- JS Metadata API.
- A lot of internal changes and bugfixes.

## Breaking changes

> * [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names incompatible with OpenTelemetry or Prometheus. This is planned to become an error with v0.48.0. 

Reviewer 3: We should also update our docs with the allowed pattern and link here. (duplicating the comment since it was the part of the previous suggestion)

Author: The issue is already marked (by you) with a documentation needed. Do you want us to expand on this in the comment?

Reviewer 3: I'd prefer to keep the comment till the actual moment when we refer here to the docs, otherwise, it could be missed.

> * [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names incompatible with OpenTelemetry or Prometheus. This is planned to [become an error](https://github.com/grafana/k6/issues/3065) with v0.48.0.

Author: ```suggestion
* [#3066](https://github.com/grafana/k6/pull/3066) k6 will warn users about metric names [incompatible with OpenTelemetry or Prometheus](https://k6.io/docs/using-k6/metrics/#metric-name-restrictions). This is planned to [become an error](https://github.com/grafana/k6/issues/3065) with v0.48.0.
```

## New features

> ### Experimental gRPC module with stream supports [#3107](https://github.com/grafana/k6/pull/3107)

Reviewer 3: ```suggestion
### Experimental gRPC module with streaming support [#3107](https://github.com/grafana/k6/pull/3107)
```

</details>

You can just replace `k6/grpc` import with `k6/experimental/grpc` to use the new functionality.

In the future this functionality will be moved to the `k6/grpc` module.

### You can now only upload a test to the cloud without running it [#3030](https://github.com/grafana/k6/pull/3030)

For [years users have wanted](https://github.com/grafana/k6-cloud-feature-requests/issues/22) to be able to update the test that is saved in the cloud but *not* run it at this exact point.

This is now possible by adding `--upload-only` when invoking `k6 cloud` as in `k6 cloud --upload-only script.js`.

This is likely going to be most useful in a CI on the actual test script project. Now that CI can just run `k6 cloud --upload-only new-verson-of-script.js` on "release".

And later on that newer version will be used. For example by a [scheduled run](https://k6.io/docs/cloud/manage/scheduled-tests/). 

### Setting sample metadata API [#3037](https://github.com/grafana/k6/pull/3037)

It is now possible to set metadata for samples emitted by a given VU. 

Metadata was added in v0.41.0 but was not available to be set from JS code. This was mainly due to lack of agreement on the API to do so.

> After further discussion a compromised smaller and consistent API was aggread upon.

Reviewer 3: ```suggestion
After further discussion, a compromised, smaller and consistent API was agreed upon.
```

> After further discussion, a compromised, smaller and consistent API was agreed upon.

Author: ```suggestion
After further discussion, a compromise for a smaller and consistent API was reached.
```

> After further discussion, a compromised, smaller and consistent API was agreed upon.

Reviewer 4: Your suggestion sounds better, but I wouldn't mention any discussions or compromises, as it's not relevant for the user.

So I would leave this as:

> Support for high-cardinality metrics metadata was added in v0.41.0, but it wasn't accessible from test scripts. It's now possible to set or delete metadata with a similar API as used for tags:

> After further discussion, a compromised, smaller and consistent API was agreed upon.

Author: Can you make it a `suggestion` so I can just merge it, please

The API is similar to the one used to set tags for all metrics of the vu:
```javascript
import exec from "k6/execution";

export default () => {
    exec.vu.metrics.metadata["my_cool_id"] = "a very unique value"
    // all metrics from here on will have this metadata set
    delete exec.vu.metrics.metadata["my_cool_id"]
    // all metriccs from here on wil *not* have the metadata set
}
```

> This also introduces the sub object `metrics` on the `vu` object.

Reviewer 3: ```suggestion
This also introduces the sub-object `metrics` on the `vu` object.
```

Apart from `metadata` it has another property `tags`. This is supposed to be the new way to set tags instead of usng `exec.vu.tags`. 

> There are no current plans when or if `exec.vu.tags` will be dropped and only `exec.vu.metrics.tags` to stay.

Author: That was the original agreement from the issue and when the API was discussed. But that is mostly relevant for the doc changes

### UX improvements and enhancements

> - [#3015](https://github.com/grafana/k6/pull/3015) Official arm64 docker images. Thanks to @nrxr! 🙇

Author: There is quite a lot of writing most of which is some technical details. 

But if there is no benefit from the changes shall we revert? If there are benefits we can mention them

> - [#3015](https://github.com/grafana/k6/pull/3015) Official arm64 docker images. Thanks to @nrxr! 🙇

Reviewer 4: There are no practical benefits for users. We build the ARM image in CI, but it's not pushed anywhere.

We don't need to revert it, since there's no harm in leaving it as is, but I'm hoping that the author can address the push issue for the next release, and we can mention this then.

> - [#3015](https://github.com/grafana/k6/pull/3015) Official arm64 docker images. Thanks to @nrxr! 🙇

Author: ```suggestion
- [#3015](https://github.com/grafana/k6/pull/3015) We started building ARM64 Docker images. There is still not pushed in the official repo though. Thanks to @nrxr! 🙇
```

> - [#3015](https://github.com/grafana/k6/pull/3015) Official arm64 docker images. Thanks to @nrxr! 🙇

Reviewer 4: I don't think we should mention it at all... There's literally no way for people to benefit from this, and it will only raise questions about how to use it. I would like to credit the author for the work done so far as well, but it doesn't make sense until the work is actually finished.

> - [#3015](https://github.com/grafana/k6/pull/3015) Official arm64 docker images. Thanks to @nrxr! 🙇

Reviewer 4: On second thought, it probably deserves to be mentioned. To credit the author, and to point out that we're working on it, even though it's not usable yet. So maybe:

```suggestion
- [#3015](https://github.com/grafana/k6/pull/3015) We're now building ARM64 Docker images in CI, but due to some Docker complications, they're still not pushed to the repository. We expect to resolve these issues by next release. Thanks @nrxr! 🙇
```

- [#3099](https://github.com/grafana/k6/pull/3099) replace "breached" with "crossed" in logs around thresholds. Thanks to @MattDodsonEnglish 🙇.
- [#3102](https://github.com/grafana/k6/pull/3102) Better error message when SharedArray constructor is provided with an async function. This is not supported, but the original message wasn't very clear.

> - [#3089](https://github.com/grafana/k6/pull/3089) Add SBOM reports to k6 releases. Thanks to @SadFaceSmith 🙇.

Reviewer 3: ```suggestion
- [#3089](https://github.com/grafana/k6/pull/3089) Add Software Bill of Materials (SBOM) reports to k6 releases. Thanks to @SadFaceSmith 🙇.
```

> - [goja#510](https://github.com/dop251/goja/pull/510) `JSON.parse` will now fail with a friendler error message.

Reviewer 4: Fixing my own typo :sweat_smile:

```suggestion
- [goja#510](https://github.com/dop251/goja/pull/510) `JSON.parse` will now fail with a friendlier error message.
```

## Maintenance and internal improvements

> - [#2091](https://github.com/grafana/k6/pull/2991) Refactor JS modules system so that is usable in tests. Which allowed enabling the tc39 tests for modules [#3040](https://github.com/grafana/k6/pull/3040).

Reviewer 4: ```suggestion
- [#2991](https://github.com/grafana/k6/pull/2991) Refactor JS modules system so that is usable in tests. Which allowed enabling the tc39 tests for modules [#3040](https://github.com/grafana/k6/pull/3040).
```

- [#3025](https://github.com/grafana/k6/pull/3025) Internally stop referring to afero and use an internal package to do all file system interaction. That package still uses afero.
- [#3036](https://github.com/grafana/k6/pull/3036) and [#3053](https://github.com/grafana/k6/pull/3053) Add options to `scenarios` for usage by browser module. 
- [#3064](https://github.com/grafana/k6/pull/3064), [#3070](https://github.com/grafana/k6/pull/3070),  [#3075](https://github.com/grafana/k6/pull/3075) and [#3106](https://github.com/grafana/k6/pull/3106) Go dependancies updates. 
- [#3067](https://github.com/grafana/k6/pull/3067) Add method to retrive all registered metrics.
- [#3068](https://github.com/grafana/k6/pull/3068) Add metric Sink constructor.
- [#3078](https://github.com/grafana/k6/pull/3078) Pinning to alpine 3.17 in docker builder image. Thank you, @arukiidou 🙇.
- [#3086](https://github.com/grafana/k6/pull/3086) Fix downlonad `.golangci.yml` for PRs from forks. 
- [#3088](https://github.com/grafana/k6/pull/3088) Make TestDNSResolver less flaky. 
- [#3094](https://github.com/grafana/k6/pull/3094) Fix example from the run command. Thanks to @rocktwotj 🙇.

> - [#3095](https://github.com/grafana/k6/pull/3095) Maintenance update of .`golangci.yml`.

Reviewer 3: ```suggestion
- [#3095](https://github.com/grafana/k6/pull/3095) Maintenance update of `.golangci.yml`.
```

- [#3103](https://github.com/grafana/k6/pull/3103) Fix lint and logical issues in `k6/data` module tests.
- [#3045](https://github.com/grafana/k6/pull/3045), [#3049](https://github.com/grafana/k6/pull/3049), [#3073](https://github.com/grafana/k6/pull/3073) and [#3044](https://github.com/grafana/k6/pull/3044) New issues are now assigned to maintaners. This is to better split the triaging of new issues between maitainers. Also, both that and new PRs assignement actions are now not using external actions.
## Roadmap

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028)

Author: I mean we are hitting the short-term goals I guess :shrug: 

Don't know if we should update them now or keep it for a little while longer :shrug: 

I guess we will need to update them each release/planning phase or something similar - so we can do it during this cooldown period

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028)

Reviewer 3: >I mean we are hitting the short-term goals I guess 

That's for sure and great :sweat_smile: 

My concern is that some people will see this doc the first time (since the release notes are quite a strong channel of promotion), and they could be confused if they see obsolete things.

So yeah this cooldown sounds great.

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028)

Reviewer: What is the plan here? Are we going to open a PR for updating it?

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028)

Reviewer: I've created a PR for it https://github.com/grafana/k6/pull/3119

The idea is that we will keep it updated and current or future users will be able to look at it to see on what bigger projects the team is working on.

### Cloud output v2

>  TODO

Reviewer: We should not mention it, we are not adding any improvement for the end user, so if we really want to mention them then let's add it to `Maintenance and internal improvements` section.

Author: This is in the future/roadmap section. And it is TODO specifically so we discuss this.

Author: Or do you mean that even when we release it - we should not be putting it in the release notes?

Reviewer: >This is in the future/roadmap section. And it is TODO specifically so we discuss this.

So would you like to announce it as an upcoming improvement? Why do you think is it important?

>Or do you mean that even when we release it - we should not be putting it in the release notes?

No, I meant in the current release notes. I expect it will be one more item in the `Maintenance and internal improvements` section when we will release it.

Author: > Why do you think is it important?

You did do quite a lot of work on it, and it will be quite a big change even if (hopefully) no users manages to notice it.

Reviewer 4: Agree with Mihail. I think this is important to mention, even though it's not quite ready for public release. But k6 Cloud users will be interested to know about the changes.

So maybe we don't need to link each individual GH issue, but mention that we're working on a Cloud output overhaul, describing all the improvements, and that it will be ready by next release.

> ## Roadmap

Reviewer: ```suggestion

## Roadmap
```

> There is now a [public Roadmap document](https://github.com/grafana/k6/blob/master/ROADMAP.md) in the main repo added in [#3028](https://github.com/grafana/k6/pull/3028).

Reviewer: ```suggestion
There is now a [public Roadmap](https://github.com/orgs/grafana/projects/443/views/1) as GitHub project of the main repo.
```

While functionally it is now mostly complete, we feel like more testing is still needed and some smaller issues need to be ironed out.

Over the next cycle we will be testing it internally, and in v0.46.0 it will be generally available as the default Cloud output. It will still be possible to use the current version via an option, but we plan to gradually deprecate it.

The new output has some benefits over the previous one:
- Binary (protobuf) format instead of JSON [#2963](https://github.com/grafana/k6/pull/2963)
- Samples aggregation for every metric instead of only for HTTP ones [#3071](https://github.com/grafana/k6/pull/3071)
- HDR Histogram generation for trend-type metrics [#3027](https://github.com/grafana/k6/pull/3027)

This in general makes the payload sent for tests with a lot of samples much smaller, which also in most cases has turned out to lower the CPU and memory usage.

>  Other related PRs: [#3041](https://github.com/grafana/k6/pull/3041), [#3061](https://github.com/grafana/k6/pull/3061), [#3063](https://github.com/grafana/k6/pull/3063) [#3072](https://github.com/grafana/k6/pull/3072), [#3082](https://github.com/grafana/k6/pull/3082), [#3083](https://github.com/grafana/k6/pull/3083), [#3085](https://github.com/grafana/k6/pull/3085), [#3094](https://github.com/grafana/k6/pull/3098), [#3105](https://github.com/grafana/k6/pull/3105)

Reviewer 4: ```suggestion
 Other related PRs: [#3041](https://github.com/grafana/k6/pull/3041), [#3061](https://github.com/grafana/k6/pull/3061), [#3063](https://github.com/grafana/k6/pull/3063), [#3072](https://github.com/grafana/k6/pull/3072), [#3082](https://github.com/grafana/k6/pull/3082), [#3083](https://github.com/grafana/k6/pull/3083), [#3085](https://github.com/grafana/k6/pull/3085), [#3098](https://github.com/grafana/k6/pull/3098), [#3105](https://github.com/grafana/k6/pull/3105)
```

And later on that newer version will be used. For example by a [scheduled run](https://k6.io/docs/cloud/manage/scheduled-tests/). 

### Setting sample metadata API [#3037](https://github.com/grafana/k6/pull/3037)

Support for high-cardinality metrics metadata was added in v0.41.0, but it wasn't accessible from test scripts. It's now possible to set or delete metadata for the whole VU with a similar API as used for tags:
```javascript
import exec from "k6/execution";

export default () => {
  exec.vu.metrics.metadata["my_cool_id"] = "a very unique value";
  // all metrics from here on will have this metadata set
  delete exec.vu.metrics.metadata["my_cool_id"];
  // all metrics from here on will *not* have the metadata set
}
```

This also introduces the sub-object `metrics` on the `vu` object.
Apart from `metadata` it has another property `tags`. This is meant to be the new way to set tags instead of using `exec.vu.tags`. 

There are no current plans to replace `exec.vu.tags` with `exec.vu.metrics.tags`.

### UX improvements and enhancements

> - [#3015](https://github.com/grafana/k6/pull/3015) We're now building ARM64 Docker images in CI, but due to some Docker complications, they're still not pushed to the repository. We expect to resolve these issues by next release. Thanks @nrxr! 🙇

Reviewer 3: ```suggestion
```

Because of https://github.com/grafana/k6/pull/3128

- [#3078](https://github.com/grafana/k6/pull/3078) Pin base Docker builder image to Alpine 3.17. Thank you, @arukiidou 🙇.
- [#3086](https://github.com/grafana/k6/pull/3086) Fix downloading `.golangci.yml` for PRs from forks. 
- [#3088](https://github.com/grafana/k6/pull/3088) Make TestDNSResolver less flaky. 
- [#3094](https://github.com/grafana/k6/pull/3094) Fix example from the run command. Thanks to @rocktwotj 🙇.
- [#3095](https://github.com/grafana/k6/pull/3095) Maintenance update of `.golangci.yml`.
- [#3103](https://github.com/grafana/k6/pull/3103) Fix lint and logical issues in `k6/data` module tests.
- [#3045](https://github.com/grafana/k6/pull/3045), [#3049](https://github.com/grafana/k6/pull/3049), [#3073](https://github.com/grafana/k6/pull/3073) and [#3044](https://github.com/grafana/k6/pull/3044) New issues are now automatically assigned to maintainers, to improve response time on issues. Both new issue and new PR assignments are now not using external actions.
- [#3109](https://github.com/grafana/k6/pull/3109) Add a way to get the cloudapi Client's base URL. Thanks to @yorugac 🙇.

## Roadmap

> There is now a [public Roadmap](https://github.com/orgs/grafana/projects/443/views/1) as a GitHub project of the main repo.
> 
> The idea is that we will keep it updated and current or future users will be able to look at it to see on what bigger projects the team is working on.

Reviewer 3: We can re-use Théo's original sentences, they sound great

```suggestion
We're excited to share [our public roadmap](https://github.com/orgs/grafana/projects/443/views/1), outlining the upcoming features and improvements we have planned.

We hope this updated roadmap provides a clear overview of our plans for k6's future development. As always, we welcome feedback, corrections, and suggestions to make this roadmap more comprehensive, accessible, and valuable for the k6 community.
```
